src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumClient.kt

```diff
@@ -12,237 +12,283 @@ import kotlinx.coroutines.flow.*
         if (_connectionStatus.value is ElectrumConnectionStatus.Closed) {
-            runJob = establishConnection(serverAddress)
-        } else logger.warning { "electrum client is already running" }
+            listenJob?.cancel()
+            val socket = openSocket(serverAddress, socketBuilder, timeout) ?: return false
+            logger.info { "connected to electrumx instance" }
+            return try {
+                handshake(socket, timeout)
+                listenJob = listen(socket)
+                true
+            } catch (ex: Throwable) {
+                logger.warning(ex) { "electrum connection handshake failed: " }
+                val ioException = when (ex) {
+                    is TcpSocket.IOException -> ex
+                    else -> TcpSocket.IOException.Unknown(ex.message, ex)
+                }
+                socket.close()
+                _connectionStatus.value = ElectrumConnectionStatus.Closed(ioException)
+                false
+            }
+        } else {
+            logger.warning { "ignoring connection request, electrum client is already running" }
+            return false
+        }
     }
 
-    @OptIn(DelicateCoroutinesApi::class)
     fun disconnect() {
-        launch {
-            if (!mailbox.isClosedForSend) {
-                mailbox.send(Action.Disconnect)
-            }
-        }
+        listenJob?.cancel()
+        listenJob = null
+        _connectionStatus.value = ElectrumConnectionStatus.Closed(null)
     }
 
-    private fun establishConnection(serverAddress: ServerAddress) = launch(CoroutineExceptionHandler { _, exception ->
-        logger.error(exception) { "error starting electrum client: " }
-    }) {
-        _connectionStatus.value = ElectrumConnectionStatus.Connecting
-        val socket: TcpSocket = try {
+    private suspend fun openSocket(serverAddress: ServerAddress, socketBuilder: TcpSocket.Builder, timeout: Duration): TcpSocket? {
+        var socket: TcpSocket? = null
+        return try {
+            _connectionStatus.value = ElectrumConnectionStatus.Connecting
             val (host, port, tls) = serverAddress
             logger.info { "attempting connection to electrumx instance [host=$host, port=$port, tls=$tls]" }
-            socketBuilder?.connect(host, port, tls, loggerFactory) ?: error("socket builder is null.")
+            withTimeout(timeout) {
+                socket = socketBuilder.connect(host, port, tls, loggerFactory)
+                socket
+            }
         } catch (ex: Throwable) {
-            logger.warning(ex) { "TCP connect: ${ex.message}: " }
+            logger.warning(ex) { "could not connect to electrum server: " }
             val ioException = when (ex) {
                 is TcpSocket.IOException -> ex
                 else -> TcpSocket.IOException.ConnectionRefused(ex)
             }
+            socket?.close()
```

Comment:
You actually need the socket to make sure you don't leak resources, it needs to be properly closed. In practice this shouldn't have much impact, but since we can easily avoid it it feels cleaner? See [this comment](https://github.com/ACINQ/lightning-kmp/pull/512#discussion_r1285613388) for the complete discussion.